store/cachekvstore.go

```diff
@@ -0,0 +1,172 @@
+package store
+
+import (
+	"bytes"
+	"sort"
+	"sync"
+)
+
+// If value is nil but deleted is false, it means the parent doesn't have the
+// key.  (No need to delete upon Write())
+type cValue struct {
+	value   []byte
+	deleted bool
+	dirty   bool
+}
+
+// cacheKVStore wraps an in-memory cache around an underlying KVStore.
+type cacheKVStore struct {
+	mtx    sync.Mutex
+	cache  map[string]cValue
+	parent KVStore
+}
+
+var _ CacheKVStore = (*cacheKVStore)(nil)
+
+func NewCacheKVStore(parent KVStore) *cacheKVStore {
+
+	ci := &cacheKVStore{
+		cache:  make(map[string]cValue),
+		parent: parent,
+	}
+
+	return ci
+}
+
+func (ci *cacheKVStore) Get(key []byte) (value []byte) {
+	ci.mtx.Lock()
+	defer ci.mtx.Unlock()
+	ci.assertValidKey(key)
+
+	cacheValue, ok := ci.cache[string(key)]
+	if !ok {
+		value = ci.parent.Get(key)
+		ci.setCacheValue(key, value, false, false)
+	} else {
+		value = cacheValue.value
+	}
+
+	return value
+}
+
+func (ci *cacheKVStore) Set(key []byte, value []byte) {
+	ci.mtx.Lock()
+	defer ci.mtx.Unlock()
+	ci.assertValidKey(key)
+
+	ci.setCacheValue(key, value, false, true)
+}
+
+func (ci *cacheKVStore) Has(key []byte) bool {
+	value := ci.Get(key)
+	return value != nil
+}
+
+func (ci *cacheKVStore) Delete(key []byte) {
+	ci.mtx.Lock()
+	defer ci.mtx.Unlock()
+	ci.assertValidKey(key)
+
+	ci.setCacheValue(key, nil, true, true)
+}
+
+// Write writes pending updates to the parent database and clears the cache.
+func (ci *cacheKVStore) Write() {
+	ci.mtx.Lock()
+	defer ci.mtx.Unlock()
+
+	// We need a copy of all of the keys.
+	// Not the best, but probably not a bottleneck depending.
+	keys := make([]string, 0, len(ci.cache))
+	for key, dbValue := range ci.cache {
+		if dbValue.dirty {
+			keys = append(keys, key)
+		}
+	}
+	sort.Strings(keys)
+
+	// TODO in tmlibs/db we use Batch to write atomically.
+	// Consider allowing usage of Batch.
+	for _, key := range keys {
+		cacheValue := ci.cache[key]
+		if cacheValue.deleted {
+			ci.parent.Delete([]byte(key))
+		} else if cacheValue.value == nil {
+			// Skip, it already doesn't exist in parent.
+		} else {
+			ci.parent.Set([]byte(key), cacheValue.value)
+		}
+	}
+
+	// Clear the cache
+	ci.cache = make(map[string]cValue)
```

Comment:
Minor nit, this finna be pressure on the garbage collector. In the future, it'd be nice to
implement
```go
for key := range ci.cache {
   delete(ci.cache, key)
}
```
in the next Go cycles, we'll be making the compiler recognize the map clearing idiom above, as per https://github.com/golang/go/issues/20138